File: Algorithms/Python/0079_Word_Search/Word_Search_Solution_1.py

# Space: O(l)
# Time: O(m * n * l)

class Solution:
    def exist(self, board, word) -> bool:

        column_length = len(board)
        row_length = len(board[0])
        word_length = len(word)

        def dfs(x, y, index):
            if index >= word_length: return True
            if (not 0 <= x < row_length) or (not 0 <= y < column_length): return False
            if board[y][x] != word[index]: return False

            temp = board[y][x]
            board[y][x] = '*'

            if dfs(x - 1, y, index + 1): return True
            if dfs(x + 1, y, index + 1): return True
            if dfs(x, y - 1, index + 1): return True
            if dfs(x, y + 1, index + 1): return True

            board[y][x] = temp

        for column in range(column_length):
            for row in range(row_length):
                if dfs(row, column, 0): return True
        return False

File: Algorithms/0079_Word_Search/Java/WordSearchSolution1.java

// Space: O(l)
// Time: O(m * n * l)

class Solution {
    public boolean exist(char[][] board, String word) {
        for (int y = 0; y < board.length; y++) {
            for (int x = 0; x < board[0].length; x++) {
                if (dfs(board, word, 0, y, x)) return true;
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, String word, int index, int y, int x) {
        if (index >= word.length()) return true;
        if (y < 0 || x < 0 || y == board.length || x == board[0].length) return false;
        if (board[y][x] != word.charAt(index)) return false;
        char temp = board[y][x];
        board[y][x] = '*';
        if (dfs(board, word, index + 1, y + 1, x)) return true;
        if (dfs(board, word, index + 1, y - 1, x)) return true;
        if (dfs(board, word, index + 1, y, x + 1)) return true;
        if (dfs(board, word, index + 1, y, x - 1)) return true;
        board[y][x] = temp;
        return false;
    }
}




